File: src/Sample/Plumix.Sample/Demos/Material/TextFieldDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/material/text_field_demo_page.dart

public sealed class TextFieldDemoPage : StatefulWidget
{
    public override State CreateState() => new TextFieldDemoPageState();

    private sealed class TextFieldDemoPageState : State
    {
        private readonly TextEditingController _email = new();
        private readonly TextEditingController _password = new();
        private readonly TextEditingController _notes = new();
        private readonly TextEditingController _readOnly = new("Read-only value");
        private readonly TextEditingController _formName = new();
        private readonly LabeledGlobalKey<FormState> _formKey = new("text-form");
        private bool _enabled = true;
        private bool _obscure = true;
        private bool _error;
        private string _submitted = "none";
        private string _formStatus = "not validated";

        public override Widget Build(BuildContext context) => new SingleChildScrollView(
            child: new Column(
                mainAxisSize: MainAxisSize.Min,
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 12,
                children:
                [
                    new Text("InputDecorator + TextField", fontSize: 20),
                    new Text(
                        "Filled/outlined/state-aware borders, floating labels, hint/helper/error/counter slots, "
                        + "prefix/suffix icons, pointer selection, adaptive context menus, read-only and "
                        + "multiline input.",
                        fontSize: 14, color: Colors.DimGray),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            Control(_enabled ? "Enabled" : "Disabled", () => SetState(() => _enabled = !_enabled)),
                            Control(_obscure ? "Reveal" : "Hide", () => SetState(() => _obscure = !_obscure)),
                            Control(_error ? "Clear error" : "Show error", () => SetState(() => _error = !_error)),
                        ]),
                    new TextField(
                        controller: _email,
                        enabled: _enabled,
                        maxLength: 32,
                        decoration: new InputDecoration(
                            labelText: "Email",
                            hintText: "name@example.com",
                            helperText: "Filled Material field",
                            prefixIcon: new Icon(Icons.Email),
                            suffixText: ".com",
                            filled: true),
                        onSubmitted: value => SetState(() => _submitted = value)),
                    new TextField(
                        controller: _password,
                        enabled: _enabled,
                        obscureText: _obscure,
                        decoration: new InputDecoration(
                            labelText: "Password",
                            errorText: _error ? "At least 8 characters" : null,
                            prefixIcon: new Icon(Icons.Lock),
                            suffixIcon: new Icon(_obscure ? Icons.Visibility : Icons.VisibilityOff),
                            border: new OutlineInputBorder(borderRadius: BorderRadius.Circular(12)))),
                    new TextField(
                        controller: _notes,
                        enabled: _enabled,
                        minLines: 3,
                        maxLines: 3,
                        decoration: new InputDecoration(
                            labelText: "Notes",
                            alignLabelWithHint: true,
                            border: new OutlineInputBorder(),
                            helperText: "Multiline EditableText path")),
                    new TextField(
                        enabled: _enabled,
                        decoration: new InputDecoration(
                            labelText: "State-aware border",
                            hintText: "Focus or hover this field",
                            errorText: _error ? "Error state" : null,
                            border: MaterialStateOutlineInputBorder.ResolveWith(ResolveStateBorder),
                            helperText: "Resolves focus, hover, error, and disabled together")),
                    new TextField(
                        controller: _readOnly,
                        readOnly: true,
                        decoration: InputDecoration.Collapsed("Read only")),
                    new Text("Decorator geometry probes", fontSize: 18),
                    new TextField(
                        enabled: _enabled,
                        decoration: new InputDecoration(
                            labelText: "Centered floating label",
                            hintText: "Gap follows the label",
                            floatingLabelAlignment: FloatingLabelAlignment.Center,
                            floatingLabelBehavior: FloatingLabelBehavior.Always,
                            border: new OutlineInputBorder())),
                    new TextField(
                        enabled: _enabled,
                        decoration: new InputDecoration(
                            labelText: "Shaped border",
                            hintText: "Any ShapeBorder as the input outline",
                            helperText: "ShapedInputBorder cuts the label gap out of the shape",
                            border: new ShapedInputBorder(
                                new StadiumBorder(),
                                new BorderSide(Colors.Indigo, 2.0)))),
                    new TextField(
                        enabled: _enabled,
                        decoration: new InputDecoration(
                            labelText: "Dense + compact density",
                            helperText: "isDense with VisualDensity.Compact",
                            isDense: true,
                            visualDensity: VisualDensity.Compact,
                            filled: true)),
                    new SizedBox(
                        height: 96,
                        child: new TextField(
                            enabled: _enabled,
                            expands: true,
                            maxLines: null,
                            textAlignVertical: TextAlignVertical.Bottom,
                            decoration: new InputDecoration(
                                labelText: "Expanded, bottom aligned",
                                filled: true))),
                    new Text($"Last submitted email: {_submitted}", fontSize: 13),
                    new Divider(),
                    new Text("TextFormField + Form", fontSize: 18),
                    new Form(
                        key: _formKey,
                        autovalidateMode: AutovalidateMode.OnUserInteraction,
                        child: new Column(
                            crossAxisAlignment: CrossAxisAlignment.Stretch,
                            spacing: 8,
                            children:
                            [
                                new TextFormField(
                                    controller: _formName,
                                    decoration: new InputDecoration(
                                        labelText: "Display name",
                                        helperText: "Required form field",
                                        border: new OutlineInputBorder()),
                                    validator: value => string.IsNullOrWhiteSpace(value)
                                        ? "Enter a display name"
                                        : value.Length < 3 ? "Use at least 3 characters" : null,
                                    onSaved: value => SetState(() => _formStatus = $"saved: {value}")),
                                new Row(
                                    spacing: 8,
                                    children:
                                    [
                                        Control("Validate", ValidateForm),
                                        Control("Save", SaveForm),
                                        Control("Reset", ResetForm),
                                    ]),
                                new Text($"Form status: {_formStatus}", fontSize: 13),
                            ])),
                ]));

        public override void Dispose()
        {
            _email.Dispose(); _password.Dispose(); _notes.Dispose(); _readOnly.Dispose(); _formName.Dispose();
        }

        private void ValidateForm()
        {
            bool valid = _formKey.CurrentState?.Validate() == true;
            SetState(() => _formStatus = valid ? "valid" : "invalid");
        }

        private void SaveForm()
        {
            var form = _formKey.CurrentState;
            if (form?.Validate() != true)
            {
                SetState(() => _formStatus = "invalid");
                return;
            }
            form.Save();
        }

        private void ResetForm()
        {
            _formKey.CurrentState?.Reset();
            SetState(() => _formStatus = "reset");
        }

        private static InputBorder ResolveStateBorder(MaterialState states)
        {
            Color color = states.HasFlag(MaterialState.Disabled)
                ? Colors.Gray
                : states.HasFlag(MaterialState.Error)
                    ? Colors.Crimson
                    : states.HasFlag(MaterialState.Focused)
                        ? Colors.DodgerBlue
                        : states.HasFlag(MaterialState.Hovered)
                            ? Colors.MediumSeaGreen
                            : Colors.SlateGray;
            double width = states.HasFlag(MaterialState.Focused) ? 3 : 1;
            return new OutlineInputBorder(new BorderSide(color, width), BorderRadius.Circular(10));
        }

        private static Widget Control(string label, Action action) => new TextButton(new Text(label, fontSize: 12), action);
    }
}

File: dart_sample/lib/demos/material/text_field_demo_page.dart
import 'package:material_ui/material_ui.dart';

class TextFieldDemoPage extends StatefulWidget {
  const TextFieldDemoPage({super.key});

  @override
  State<TextFieldDemoPage> createState() => _TextFieldDemoPageState();
}

class _TextFieldDemoPageState extends State<TextFieldDemoPage> {
  final TextEditingController _email = TextEditingController();
  final TextEditingController _password = TextEditingController();
  final TextEditingController _notes = TextEditingController();
  final TextEditingController _readOnly = TextEditingController(
    text: 'Read-only value',
  );
  final TextEditingController _formName = TextEditingController();
  final GlobalKey<FormState> _formKey = GlobalKey<FormState>();
  bool _enabled = true;
  bool _obscure = true;
  bool _error = false;
  String _submitted = 'none';
  String _formStatus = 'not validated';

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 12,
        children: <Widget>[
          const Text(
            'InputDecorator + TextField',
            style: TextStyle(fontSize: 20),
          ),
          const Text(
            'Filled/outlined/state-aware borders, floating labels, hint/helper/error/counter slots, '
            'prefix/suffix icons, pointer selection, adaptive context menus, read-only and multiline input.',
            style: TextStyle(fontSize: 14, color: Colors.black54),
          ),
          Wrap(
            spacing: 8,
            runSpacing: 8,
            children: <Widget>[
              TextButton(
                onPressed: () => setState(() => _enabled = !_enabled),
                child: Text(_enabled ? 'Enabled' : 'Disabled'),
              ),
              TextButton(
                onPressed: () => setState(() => _obscure = !_obscure),
                child: Text(_obscure ? 'Reveal' : 'Hide'),
              ),
              TextButton(
                onPressed: () => setState(() => _error = !_error),
                child: Text(_error ? 'Clear error' : 'Show error'),
              ),
            ],
          ),
          TextField(
            controller: _email,
            enabled: _enabled,
            maxLength: 32,
            decoration: const InputDecoration(
              labelText: 'Email',
              hintText: 'name@example.com',
              helperText: 'Filled Material field',
              prefixIcon: Icon(Icons.email),
              suffixText: '.com',
              filled: true,
            ),
            onSubmitted: (String value) => setState(() => _submitted = value),
          ),
          TextField(
            controller: _password,
            enabled: _enabled,
            obscureText: _obscure,
            decoration: InputDecoration(
              labelText: 'Password',
              errorText: _error ? 'At least 8 characters' : null,
              prefixIcon: const Icon(Icons.lock),
              suffixIcon: Icon(
                _obscure ? Icons.visibility : Icons.visibility_off,
              ),
              border: OutlineInputBorder(
                borderRadius: BorderRadius.circular(12),
              ),
            ),
          ),
          TextField(
            controller: _notes,
            enabled: _enabled,
            minLines: 3,
            maxLines: 3,
            decoration: const InputDecoration(
              labelText: 'Notes',
              alignLabelWithHint: true,
              border: OutlineInputBorder(),
              helperText: 'Multiline EditableText path',
            ),
          ),
          TextField(
            enabled: _enabled,
            decoration: InputDecoration(
              labelText: 'State-aware border',
              hintText: 'Focus or hover this field',
              errorText: _error ? 'Error state' : null,
              border: WidgetStateInputBorder.resolveWith(_resolveStateBorder),
              helperText: 'Resolves focus, hover, error, and disabled together',
            ),
          ),
          TextField(
            controller: _readOnly,
            readOnly: true,
            decoration: const InputDecoration.collapsed(hintText: 'Read only'),
          ),
          const Text('Decorator geometry probes', style: TextStyle(fontSize: 18)),
          TextField(
            enabled: _enabled,
            decoration: const InputDecoration(
              labelText: 'Centered floating label',
              hintText: 'Gap follows the label',
              floatingLabelAlignment: FloatingLabelAlignment.center,
              floatingLabelBehavior: FloatingLabelBehavior.always,
              border: OutlineInputBorder(),
            ),
          ),
          TextField(
            enabled: _enabled,
            decoration: const InputDecoration(
              labelText: 'Shaped border',
              hintText: 'Any ShapeBorder as the input outline',
              helperText: 'ShapedInputBorder cuts the label gap out of the shape',
              border: ShapedInputBorder(
                shape: StadiumBorder(),
                borderSide: BorderSide(color: Colors.indigo, width: 2.0),
              ),
            ),
          ),
          TextField(
            enabled: _enabled,
            decoration: const InputDecoration(
              labelText: 'Dense + compact density',
              helperText: 'isDense with VisualDensity.compact',
              isDense: true,
              visualDensity: VisualDensity.compact,
              filled: true,
            ),
          ),
          SizedBox(
            height: 96,
            child: TextField(
              enabled: _enabled,
              expands: true,
              maxLines: null,
              textAlignVertical: TextAlignVertical.bottom,
              decoration: const InputDecoration(
                labelText: 'Expanded, bottom aligned',
                filled: true,
              ),
            ),
          ),
          const Text(
            'Ambient InputDecorationTheme probes',
            style: TextStyle(fontSize: 18),
          ),
          const Text(
            "The theme's fill, floating label and active indicator resolve per state, and an "
            'ambient IconButtonTheme colors the affix icons.',
            style: TextStyle(fontSize: 13, color: Colors.black54),
          ),
          InputDecorationTheme(
            data: InputDecorationThemeData(
              filled: true,
              fillColor: WidgetStateColor.resolveWith(_resolveStateFill),
              floatingLabelStyle: WidgetStateTextStyle.resolveWith(_resolveStateLabel),
              activeIndicatorBorder: WidgetStateBorderSide.resolveWith(
                _resolveStateIndicator,
              ),
            ),
            child: IconButtonTheme(
              data: IconButtonThemeData(
                style: ButtonStyle(
                  foregroundColor: WidgetStateProperty.resolveWith(
                    _resolveAffixColor,
                  ),
                ),
              ),
              child: TextField(
                enabled: _enabled,
                decoration: InputDecoration(
                  labelText: 'State-resolving theme',
                  hintText: 'Focus, hover or break this field',
                  errorText: _error ? 'Error state' : null,
                  prefixIcon: const Icon(Icons.lock),
                  suffixIcon: const Icon(Icons.visibility),
                  helperText:
                      'fillColor/floatingLabelStyle/activeIndicatorBorder',
                ),
              ),
            ),
          ),
          Text(
            'Last submitted email: $_submitted',
            style: const TextStyle(fontSize: 13),
          ),
          const Divider(),
          const Text('TextFormField + Form', style: TextStyle(fontSize: 18)),
          Form(
            key: _formKey,
            autovalidateMode: AutovalidateMode.onUserInteraction,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              spacing: 8,
              children: <Widget>[
                TextFormField(
                  controller: _formName,
                  decoration: const InputDecoration(
                    labelText: 'Display name',
                    helperText: 'Required form field',
                    border: OutlineInputBorder(),
                  ),
                  validator: (String? value) {
                    if (value == null || value.trim().isEmpty) {
                      return 'Enter a display name';
                    }
                    return value.length < 3
                        ? 'Use at least 3 characters'
                        : null;
                  },
                  onSaved: (String? value) =>
                      setState(() => _formStatus = 'saved: $value'),
                ),
                Wrap(
                  spacing: 8,
                  runSpacing: 8,
                  children: <Widget>[
                    _control('Validate', _validateForm),
                    _control('Save', _saveForm),
                    _control('Reset', _resetForm),
                  ],
                ),
                Text(
                  'Form status: $_formStatus',
                  style: const TextStyle(fontSize: 13),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  @override
  void dispose() {
    _email.dispose();
    _password.dispose();
    _notes.dispose();
    _readOnly.dispose();
    _formName.dispose();
    super.dispose();
  }

  void _validateForm() {
    final bool valid = _formKey.currentState?.validate() ?? false;
    setState(() => _formStatus = valid ? 'valid' : 'invalid');
  }

  void _saveForm() {
    final FormState? form = _formKey.currentState;
    if (form?.validate() != true) {
      setState(() => _formStatus = 'invalid');
      return;
    }
    form!.save();
  }

  void _resetForm() {
    _formKey.currentState?.reset();
    setState(() => _formStatus = 'reset');
  }

  Color _resolveStateFill(Set<WidgetState> states) {
    if (states.contains(WidgetState.disabled)) {
      return const Color(0xFFDCDCDC);
    }
    if (states.contains(WidgetState.error)) {
      return const Color(0xFFFFE4E1);
    }
    if (states.contains(WidgetState.focused)) {
      return const Color(0xFFF0F8FF);
    }
    if (states.contains(WidgetState.hovered)) {
      return const Color(0xFFF0FFF0);
    }
    return const Color(0xFFF5F5F5);
  }

  TextStyle _resolveStateLabel(Set<WidgetState> states) => TextStyle(
    color: states.contains(WidgetState.error)
        ? const Color(0xFFDC143C)
        : states.contains(WidgetState.focused)
        ? const Color(0xFF1E90FF)
        : const Color(0xFF708090),
  );

  BorderSide _resolveStateIndicator(Set<WidgetState> states) => BorderSide(
    color: states.contains(WidgetState.error)
        ? const Color(0xFFDC143C)
        : states.contains(WidgetState.focused)
        ? const Color(0xFF1E90FF)
        : states.contains(WidgetState.hovered)
        ? const Color(0xFF3CB371)
        : const Color(0xFF708090),
    width: states.contains(WidgetState.focused) ? 3 : 1,
  );

  Color _resolveAffixColor(Set<WidgetState> states) =>
      states.contains(WidgetState.error)
      ? const Color(0xFFDC143C)
      : const Color(0xFF4B0082);

  InputBorder _resolveStateBorder(Set<WidgetState> states) {
    final Color color = states.contains(WidgetState.disabled)
        ? Colors.grey
        : states.contains(WidgetState.error)
        ? Colors.red
        : states.contains(WidgetState.focused)
        ? Colors.blue
        : states.contains(WidgetState.hovered)
        ? Colors.green
        : Colors.blueGrey;
    final double width = states.contains(WidgetState.focused) ? 3 : 1;
    return OutlineInputBorder(
      borderSide: BorderSide(color: color, width: width),
      borderRadius: BorderRadius.circular(10),
    );
  }

  Widget _control(String label, VoidCallback onPressed) => TextButton(
    onPressed: onPressed,
    child: Text(label, style: const TextStyle(fontSize: 12)),
  );
}
